walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,119 @@
 
 package com.google.sps;
 
-import java.util.Collection;
+import com.google.sps.Event;
+import com.google.sps.Events;
+import com.google.sps.MeetingRequest;
+import com.google.sps.TimeRange;
+import java.util.*;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // If the request meeting is longer than the duration of one day, return an empty list.
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return Collections.emptyList();
+    }
+    // If there is no event on the calendar currently, return whole day as possible time slot.
+    if (events.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    Collection<String> mandatoryAttendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    // If no mandatory attendees or optional attendees, return whole day.
+    if (mandatoryAttendees.isEmpty() && optionalAttendees.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+    // Get unavailable time of mandatory attendees.
+    List<TimeRange> unavailableTimes = getUnavailableTimes(events, mandatoryAttendees);
+    // Merge unavailable intervals with overlaps.
+    List<TimeRange> mergedUnavailableTimes = mergeUnavailableTimes(unavailableTimes);
+    // Get available time slot base on merged unavailable time.
+    List<TimeRange> availableTimes = getAvailableTimes(mergedUnavailableTimes, request.getDuration());
+    // Deal with optional attendees.
+    if (!optionalAttendees.isEmpty()) {
+      List<TimeRange> optionalUnavailableTimes = getUnavailableTimes(events, optionalAttendees);
+      optionalUnavailableTimes.addAll(mergedUnavailableTimes);
+      List<TimeRange> optionalMergedUnavailableTimes = mergeUnavailableTimes(optionalUnavailableTimes);
+      List<TimeRange> combineAvailableTimes = getAvailableTimes(optionalMergedUnavailableTimes, request.getDuration());
```

Comment:
In these ~10 lines, there are 9 symbols containing the substring "available" used a combined total of 18 times.
- unavailableTimes
- getUnavailableTimes
- mergedUnavailableTimes
- mergeUnavailableTimes
- availableTimes
- optionalUnavailableTimes
- optionalMergedUnavailableTimes
- combineAvailableTimes

I think the naming here requires readers to spend a lot of energy focusing and reading multiple times to understand what's happening here.

I mainly want to point it out as food for thought and am not recommending a particular solution. But for example, I think this reads a little easier and is a bit more accurate as to what the methods do:
```java
List<TimeRange> unavailableTimes = findBusyTimes(events, mandatoryAttendees);
unavailableTimes = mergeOverlappingTimes(unavailableTimes);
List<TimeRange> availableTimes = findAvailableTimes(unavailableTimes, request.getDuration());
```